jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdAuthenticator.java

```diff
@@ -260,71 +272,78 @@ public Authentication validateRequest(ServletRequest req, ServletResponse res, b
+                }
 
-                        response.setContentLength(0);
-                        baseResponse.sendRedirect(nuri, true);
-                        return openIdAuth;
-                    }
+                // Attempt to login with the provided authCode.
+                OpenIdCredentials credentials = new OpenIdCredentials(authCode, getRedirectUri(request));
+                UserIdentity user = login(null, credentials, request);
+                if (user == null)
+                {
+                    sendError(request, response, null);
+                    return Authentication.SEND_FAILURE;
                 }
 
-                // Not authenticated.
-                sendError(request, response, null);
-                return Authentication.SEND_FAILURE;
+                OpenIdAuthentication openIdAuth = new OpenIdAuthentication(getAuthMethod(), user);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("authenticated {}->{}", openIdAuth, uriRedirectInfo.getUri());
+
+                // Save redirect info in session so original request can be restored after redirect.
+                synchronized (session)
+                {
+                    session.setAttribute(J_URI, uriRedirectInfo.getUri());
+                    session.setAttribute(J_METHOD, uriRedirectInfo.getMethod());
+                    session.setAttribute(J_POST, uriRedirectInfo.getFormParameters());
+                }
+
+                // Redirect to the original URI.
+                response.setContentLength(0);
+                baseResponse.sendRedirect(uriRedirectInfo.getUri(), true);
+                return openIdAuth;
             }
 
             // Look for cached authentication in the Session.
-            Authentication authentication = (Authentication)session.getAttribute(SessionAuthentication.__J_AUTHENTICATED);
-            if (authentication != null)
+            synchronized (session)
             {
-                // Has authentication been revoked?
-                if (authentication instanceof Authentication.User && _loginService != null &&
-                    !_loginService.validate(((Authentication.User)authentication).getUserIdentity()))
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("auth revoked {}", authentication);
-                    session.removeAttribute(SessionAuthentication.__J_AUTHENTICATED);
-                }
-                else
+                Authentication authentication = (Authentication)session.getAttribute(SessionAuthentication.__J_AUTHENTICATED);
+                if (authentication != null)
                 {
-                    synchronized (session)
+                    // Has authentication been revoked?
+                    if (authentication instanceof Authentication.User && _loginService != null &&
+                        !_loginService.validate(((Authentication.User)authentication).getUserIdentity()))
```

Comment:
This is a call to `LoginService` with the lock held on the `session`. I would prefer to have it outside the lock.